src/main/java/com/vmware/vipclient/i18n/messages/service/ComponentService.java

```diff
@@ -111,20 +130,42 @@ public MessageCacheItem getMessages(Iterator<Locale> fallbackLocalesIter) {
 		} else { // Item is not in cache.
 			ProductService ps = new ProductService(dto);
 			Locale locale = Locale.forLanguageTag(dto.getLocale());
-			if (ps.isSupportedLocale(locale) || VIPCfg.getInstance().isPseudo()) {
+
+			/*
+			 * Do not block createCacheItem if set of supported locales is not in cache (i.e. ps.getSupportedLocales(false).isEmpty()).
+			 * This happens either when cache is not initialized, OR previous attempts to fetch the set had failed.
+			 */
+			if (ps.getSupportedLocales(false).isEmpty() || ps.isSupportedLocale(false, locale) || VIPCfg.getInstance().isPseudo()) {
 				cacheItem = createCacheItem(); // Fetch for the requested locale from data store, create cacheItem and store in cache
 				if (cacheItem.getCachedData().isEmpty())  // Failed to fetch messages for the requested locale
 					cacheItem = getFallbackLocaleMessages(fallbackLocalesIter);
 			} else   // Requested locale is not supported and does not match any supported locales
 				cacheItem = getFallbackLocaleMessages(fallbackLocalesIter);
 		}
+
 		return cacheItem;
 	}
 
+	private void refreshSupportedLocalesTask() {
+		ProductService ps = new ProductService(dto);
+		Runnable runnable = () -> {
+			try {
+				ps.getSupportedLocales();
+			} catch (Exception e) {
+			}
+		};
+		new Thread(runnable).start();
+	}
+
 	private void doLocaleMatching() {
-		Locale matchedLocale = LocaleUtility.pickupLocaleFromList(new ProductService(dto).getSupportedLocales(), Locale.forLanguageTag(dto.getLocale()));
-		if (matchedLocale != null) // Requested locale matches a supported locale (eg. requested locale "fr_CA matches supported locale "fr")
+		//Refresh the cache of supported locales as needed in a separate thread (non-blocking).
+		refreshSupportedLocalesTask();
```

Comment:
I think before starting this task you need check if the supported locales list in cache by calling 'new ProductService(dto).getSupportedLocales(false)', if not in cache you need start this task, otherwise you needn't refresh cache.

Without the check, each call of 'getMessage' method will trigger the cache refresh of supported locale list actually which is needn't.

Reply:
If you trace this code, it will eventually call `ProductService.getSupportedLanguageTags(DataSourceEnum)`. In there, it returns the cached data if it exists in cache and is not expired.